src/tcp.cpp

```diff
@@ -151,25 +153,66 @@ int Tcp::error() const
  */
 Inbound *Tcp::send(const Query &query)
 {
-    // @todo see https://tools.ietf.org/html/rfc7766#section-7
-    // We MUST avoid having duplicate ID's that are active on the same connection,
-    // for example by delaying queries with identical ids until earlier queries are ready
-    
+    // check if this query id is inflight
+    if (_queryids.count(query.id()))
+    {
+        // it's already inflight. we'll put it in a queue to be sent later
+        auto iter = _awaiting.find(query.id());
+
+        // if there's no list of queries to be sent at this position
+        if (iter == _awaiting.end())
+        {
+            // create the list now
+            decltype(_awaiting)::mapped_type list;
+
+            // put the first element into it
+            list.push_back(query);
+
+            // update the map
+            _awaiting.emplace(query.id(), move(list));
+        }
+        else
+        {
+            // there's already a list at this key, so we can add another query to it
+            iter->second.push_back(query);
+        }
+
+        // We'll pretend everything went OK. If it later turns out we couldn't
+        // send it after all, we'll enter the `fail()` method.
+        return this;
+    }
+    else
+    {
+        // this query is not inflight, let's remember that it is in fact in flight
+        _queryids.insert(query.id());
+    }
+
+    // send it over the wire
+    return sendimpl(query);
```

Comment:
When this fails, the _querids member should not be updated.

Reply:
fixed in 203d716